app/src/org/commcare/utils/EncryptionUtils.java

```diff
@@ -1,18 +1,164 @@
 package org.commcare.utils;
 
+import android.content.Context;
+
+import org.commcare.CommCareApplication;
 import org.commcare.util.Base64;
+import org.commcare.util.Base64DecoderException;
 
+import java.io.IOException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KeyStoreException;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.SecureRandom;
+import java.security.UnrecoverableEntryException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.util.Random;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.IvParameterSpec;
 
 /**
- * Utility class for encrypting submissions during the SaveToDiskTask.
+ * Utility class for encryption functionality.
+ * Usages include:
+ * -Generating/storing/retrieving an encrypted, base64-encoded passphrase for the Connect DB
+ * -Encrypting submissions during the SaveToDiskTask.
  *
  * @author mitchellsundt@gmail.com
  */
+
 public class EncryptionUtils {
 
-    public static String getMD5HashAsString(String plainText) {
+    private static final int PASSPHRASE_LENGTH = 32;
+
+    //Generate a random passphrase
+    public static byte[] generatePassphrase() {
+        Random random;
+        try {
+            //Use SecureRandom if possible (specifying algorithm for older versions of Android)
+            random = android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O ?
+                    SecureRandom.getInstanceStrong() : SecureRandom.getInstance("SHA1PRNG");
+        } catch (NoSuchAlgorithmException e) {
+            //Fallback to basic Random
+            random = new Random();
+        }
+
+        byte[] result = new byte[PASSPHRASE_LENGTH];
+
+        while (true) {
```

Comment:
Any idea what's the average number of iterations it will take to arrive at a passphrase that doesn't contain zero ? Mostly thinking if it can result into a near infinite loop.

Reply:
On average, it will take ~1.13 iterations to generate a passphrase without any zeros, given the probability of success in each attempt (~88.4%).
For safety i have included the max attemts limit